util/compression.h

```diff
@@ -1433,51 +1437,69 @@ inline bool ZSTD_TrainDictionarySupported() {
 
 inline std::string ZSTD_TrainDictionary(const std::string& samples,
                                         const std::vector<size_t>& sample_lens,
-                                        size_t max_dict_bytes) {
+                                        size_t max_dict_bytes,
+                                        bool finalize_dict = false) {
   // Dictionary trainer is available since v0.6.1 for static linking, but not
   // available for dynamic linking until v1.1.3. For now we enable the feature
   // in v1.1.3+ only.
-#if ZSTD_VERSION_NUMBER >= 10103  // v1.1.3+
+#if defined(ZSTD) && (ZSTD_VERSION_NUMBER >= 10103)  // v1.1.3+
   assert(samples.empty() == sample_lens.empty());
   if (samples.empty()) {
     return "";
   }
+  unsigned nbSamples = static_cast<unsigned>(sample_lens.size());
   std::string dict_data(max_dict_bytes, '\0');
   size_t dict_len = ZDICT_trainFromBuffer(
-      &dict_data[0], max_dict_bytes, &samples[0], &sample_lens[0],
-      static_cast<unsigned>(sample_lens.size()));
+      &dict_data[0], max_dict_bytes, &samples[0], &sample_lens[0], nbSamples);
   if (ZDICT_isError(dict_len)) {
     return "";
   }
   assert(dict_len <= max_dict_bytes);
   dict_data.resize(dict_len);
-  return dict_data;
+  if (!finalize_dict) {
+    return dict_data;
+  }
+
+  ZDICT_params_t params;
+  memset(&params, 0, sizeof(params));
+  std::string fin_dict_data(max_dict_bytes, '\0');
+  size_t fin_dict_len = ZDICT_finalizeDictionary(
```

Comment:
Is this right? I thought ZDICT_finalizeDictionary() was an alternative to calling ZDICT_trainFromBuffer(), not an API to be called after training. See https://github.com/facebook/zstd/blob/eadb6c874f9d0c9e90c835f8b0181da802361e4c/lib/zdict.h#L148-L157